File: Src/Engine/Editor/GUI/Docking/FloatWindowDockPanel.cpp

#include "FloatWindowDockPanel.h"

#include "DockHintWindow.h"
#include "DockWindow.h"
#include "MasterDockPanel.h"
#include "Runtime/Graphics/GraphicWindow.h"
#include "Runtime/UI/GUI/WindowRootControl.h"

namespace SE::Editor
{
	void FloatWindowDockPanel::OnLeftButtonHit(WindowHitCodes hitTest, bool& result)
	{
		if (hitTest == WindowHitCodes::Caption)
		{
			BeginDrag();
			result = true;
		}
		else
		{
			result = false;
		}
	}

	void FloatWindowDockPanel::OnClosing(ClosingReason reason, bool& cancel)
	{
		// Close all docked windows
		while (Tabs.operator->().Count() > 0)
		{
			if (!Tabs.operator->()[0]->Close(reason))
			{
				// Cancel
				cancel = true;
				return;
			}
		}

		// Unlink
		_window->Window()->ClosingEvent.Unbind(_ClosingCall);;
		_window->Window()->LeftButtonHitEvent.UnbindAll();
		_window = nullptr;

		// Remove object
		ENGINE_ASSERT(TabsCount == 0 && ChildPanelsCount == 0);
		Dispose();
	}

	FloatWindowDockPanel::FloatWindowDockPanel(MasterDockPanel* masterPanel, RootControl* window) : DockPanel(nullptr)
	{
		_ClosingCall.Bind<FloatWindowDockPanel, &FloatWindowDockPanel::OnClosing>(this);
		_LeftButtonHitCall.Bind<FloatWindowDockPanel, &FloatWindowDockPanel::OnLeftButtonHit>(this);
		_masterPanel = masterPanel;
		_window = static_cast<WindowRootControl*>(window);

		// Link
		_masterPanel->FloatingPanels.Add(this);
		Parent = window;
		_window->Window()->ClosingEvent.BindUnique(_ClosingCall);;
		_window->Window()->LeftButtonHitEvent.BindUnique(_LeftButtonHitCall);
	}

	void FloatWindowDockPanel::BeginDrag()
	{
		// Filter invalid events
		if (_window == nullptr)
			return;

		// Create docking hint window
		DockHintWindow::Create(this);
	}

	GraphicWindow* FloatWindowDockPanel::CreateFloatWindow(RootControl* parent, Float2 location, Float2 size, WindowStartPosition startPosition, String title)
	{
		// Setup initial window settings
		CreateWindowSettings settings =  DefaultWindowSettings();
		WindowRootControl* t;
		if (TypeTryCast<WindowRootControl>(parent, t))
		{
			settings.Parent = t->Window();
		}
		settings.Parent = nullptr;
		settings.Title = title;
		settings.Size = size;
		settings.Position = location;
		settings.MinimumSize = Float2(1);
		settings.MaximumSize = Float2::Zero; // Unlimited size
		settings.Fullscreen = false;
		settings.HasBorder = true;
		settings.SupportsTransparency = false;
		settings.ActivateWhenFirstShown = true;
		settings.AllowInput = true;
		settings.AllowMinimize = true;
		settings.AllowMaximize = true;
		settings.AllowDragAndDrop = true;
		settings.IsTopmost = false;
		settings.IsRegularWindow = true;
		settings.HasSizingFrame = true;
		settings.ShowAfterFirstPaint = false;
		settings.ShowInTaskbar = true;
		settings.StartPosition = startPosition;

		// Create window
		return New<GraphicWindow>(settings);
	}

	DockState FloatWindowDockPanel::TryGetDockState(float& splitterValue)
	{
		splitterValue = 0.5f;
		return DockState::Float;
	}

	void FloatWindowDockPanel::OnDestroy()
	{
		if (_masterPanel != nullptr)
		{
			_masterPanel->FloatingPanels.Remove(this);
		}

		DockPanel::OnDestroy();
	}

	void FloatWindowDockPanel::OnLastTabRemoved()
	{
		// Close window
		if (_window != nullptr)
		{
			_window->Close();
		}
	}

	void FloatWindowDockPanel::OnSelectedTabChanged()
	{
		DockPanel::OnSelectedTabChanged();

		if (_window != nullptr && SelectedTab != nullptr)
			_window->SetTitle(SelectedTab->Title.operator->());
	}
} // SE

File: Src/Engine/Editor/GUI/Docking/FloatWindowDockPanel.cs
namespace SE.Editor.GUI
{
    public sealed class FloatWindowDockPanel : DockPanel
    {
        private SE.GUI.WindowRootControl? m_HostRoot;

        public FloatWindowDockPanel(MasterDockPanel masterPanel)
            : base(null)
        {
            MasterPanel = masterPanel;
        }

        public FloatWindowDockPanel(MasterDockPanel masterPanel, SE.GUI.WindowRootControl window)
            : this(masterPanel)
        {
            MasterPanel.FloatingPanels.Add(this);
            SE.Window host = window.Window;
            HostWindow = host;
            m_HostRoot = window;
            window.AddChild(this);
            m_HostRoot.SizeChanged += OnHostRootSizeChanged;
            host.Closing += OnHostWindowClosing;
            host.LeftButtonHit = OnLeftButtonHit;
        }

        public MasterDockPanel MasterPanel { get; }
        public SE.GUI.WindowRootControl? Window => m_HostRoot;
        public override bool IsFloating => true;
        public SE.Window? HostWindow { get; private set; }

        public void BeginDrag()
        {
            if (HostWindow != null)
                DockHintWindow.Create(this);
        }

        public static SE.Window CreateFloatWindow(
            SE.GUI.RootControl? parent,
            Float2 location,
            Float2 size,
            WindowStartPosition startPosition,
            string title)
        {
            // Native code currently clears the candidate parent before creating the window.
            _ = parent;
            CreateWindowSettings settings = SE.Window.CreateDefaultSettings();
            settings.Title = title;
            settings.Size = size;
            settings.Position = location;
            settings.MinimumSize = Float2.One;
            settings.MaximumSize = Float2.Zero;
            settings.Fullscreen = false;
            settings.HasBorder = true;
            settings.SupportsTransparency = false;
            settings.ActivateWhenFirstShown = true;
            settings.AllowInput = true;
            settings.AllowMinimize = true;
            settings.AllowMaximize = true;
            settings.AllowDragAndDrop = true;
            settings.IsTopmost = false;
            settings.IsRegularWindow = true;
            settings.HasSizingFrame = true;
            settings.ShowAfterFirstPaint = false;
            settings.ShowInTaskbar = true;
            settings.StartPosition = (SE.WindowStartPosition)(int)startPosition;
            return SE.Window.Create(settings);
        }

        public override DockState TryGetDockState(out float splitterValue)
        {
            splitterValue = 0.5f;
            return DockState.Float;
        }

        internal void AttachHostWindow(SE.Window window)
        {
            HostWindow = window;
            m_HostRoot = window.GUI;
            m_HostRoot.SizeChanged += OnHostRootSizeChanged;
            window.Closing += OnHostWindowClosing;
            window.LeftButtonHit = OnLeftButtonHit;
        }

        protected override void OnLastTabRemoved()
        {
            if (HostWindow != null)
                HostWindow.Close();
        }

        protected override void OnSelectedTabChanged()
        {
            base.OnSelectedTabChanged();
            if (SelectedTab != null)
            {
                if (HostWindow != null)
                    HostWindow.Title = SelectedTab.Title;
            }
        }

        protected override void OnDispose()
        {
            if (m_HostRoot != null)
            {
                m_HostRoot.SizeChanged -= OnHostRootSizeChanged;
                m_HostRoot = null;
            }

            if (HostWindow != null)
            {
                HostWindow.Closing -= OnHostWindowClosing;
                HostWindow.LeftButtonHit = null;
                HostWindow = null;
            }

            MasterPanel.FloatingPanels.Remove(this);
            base.OnDispose();
        }

        private void OnHostRootSizeChanged(SE.GUI.Control root)
        {
            SetBounds(0.0f, 0.0f, root.Width, root.Height);
        }

        private bool OnLeftButtonHit(WindowHitCodes hit)
        {
            if (hit != WindowHitCodes.Caption)
                return false;

            BeginDrag();
            return true;
        }

        private void OnHostWindowClosing(SE.ClosingReason reason, ref bool cancel)
        {
            ClosingReason dockReason = (ClosingReason)(int)reason;
            while (TabsCount > 0)
            {
                if (!Tabs[0].Close(dockReason))
                {
                    cancel = true;
                    return;
                }
            }

            SE.Window? window = HostWindow;
            if (window != null)
            {
                window.Closing -= OnHostWindowClosing;
                window.LeftButtonHit = null;
            }
            HostWindow = null;
            Dispose();
        }
    }
}
